### ISSUE TITLE
Consider adding graceful delay on termination

### ISSUE THREAD

**Issue by Benjaminvdv** [fluent/fluent-bit#6400]
**Is your feature request related to a problem? Please describe.**
When fluent-bit is deployed as a sidecar container in a Kubernetes cluster, there is a small time window where the pod received a `SIGTERM` signal and still receives traffic. This is because the cluster sends the `SIGTERM` signal and updates the loadbalancer (iptables) in parallel. This leaves a small time period where the pod is marked as `terminating` but still receives requests. 
Though this problem is not limited to fluent-bit, it impacts any multi-container pod, fluent-bit does not have any option to gracefully handle this time window.
Fluent-bit closes almost immediately if there are no pending tasks. There are two variables to be set here, `flush` and `grace` ([docs](https://docs.fluentbit.io/manual/administration/configuring-fluent-bit/yaml/configuration-file#config_section)).
`flush` sets the the frequency in which the input is flushed.
`grace` is the maximum time the service waits for finishing the flush process. if there are no tasks pending, the process immediately ends.

**Describe the solution you'd like**
I would like to suggest a `delay-shutdown` or ` shutdown-wait-period` option in the `service` configuration section. This option configures the time in seconds before shutting down after receiving `SIGTERM`. I noticed a very similar option in the [OPA repository](https://github.com/open-policy-agent/opa/issues/2764). 

**Describe alternatives you've considered**
There are a few alternatives:
* Do not use sidecars. 
* Use containers with `debug` tags. Containers published with `debug` tags are not distroless and thus have access to the `sleep` command. With this, it is possible to use `preStop` hook in Kubernetes. This alternative is described [here](https://learnk8s.io/graceful-shutdown) and [here](https://blog.sebastian-daschner.com/entries/zero-downtime-updates-kubernetes) (taken from the OPA issue mentioned above).

**Additional context**
In our use case, we are forwarding specific logs for business reports. We are not using fluent-bit for operational logs (container logs `stdout` / `stderr`). Thus running fluent-bit as a daemonset on the cluster is not desirable. Conceptually, we'd prefer the sidecar setup.

I'm very much open to new views on this issue by the fluent-bit community.

**Comment by Benjaminvdv** [comment]
Could this request receive feedback?

**Comment by JKerwood** [comment]
+1

**Comment by c4th4r5y5** [comment]
+1

**Comment by albertocavalcante** [comment]
Any update on this issue?

**Comment by leonardo-albertovich** [comment]
This sounds reasonable, I don't have the bandwidth to take care of it right now but if anyone is interested in contributing and needs guidance, feel free to contact me in slack, I'd be happy to help.

**Comment by JKerwood** [comment]
Does anyone have time to look into this.  It would be very helpful for dockerized containers.

**Comment by Berg-it** [comment]
+1

**Comment by joshsleeper** [comment]
Presto, unstale!

Sadly debug tag images are insanely big last I checked (over a gig!), so figuring out how to have graceful restarts or a native sleep command in the regular images is still high value.

**Comment by wtanner** [comment]
I ran across this issue in the process of designing a system with similar requirements. I haven't tried this within the context of a Kubernetes `preStop` lifecycle hook yet, but I was able to get the `fluent-bit` program itself to block for e.g. 30 seconds with the following input arguments:

```bash
time fluent-bit -i dummy -o exit -f 30 
Fluent Bit v2.2.1
* Copyright (C) 2015-2023 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

       \ || /
       = ** =
        /__\ 
       /____\
      /______\
     /________\
    /__________\
   /____________\
  /__Fluent Bit__\
 /__|__v2.2.1_|___\
/___\_________/____\
        |||
        |||
       =====

[2023/12/28 00:37:28] [ info] [fluent bit] version=2.2.1, commit=, pid=10626
[2023/12/28 00:37:28] [ info] [storage] ver=1.5.1, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2023/12/28 00:37:28] [ info] [cmetrics] version=0.6.6
[2023/12/28 00:37:28] [ info] [ctraces ] version=0.4.0
[2023/12/28 00:37:28] [ info] [input:dummy:dummy.0] initializing
[2023/12/28 00:37:28] [ info] [input:dummy:dummy.0] storage_strategy='memory' (memory only)
[2023/12/28 00:37:28] [ info] [sp] stream processor started
[2023/12/28 00:37:58] [ warn] [engine] service will shutdown in max 5 seconds
[2023/12/28 00:37:58] [ info] [input] pausing dummy.0
[2023/12/28 00:37:59] [ info] [engine] service has stopped (0 pending tasks)
[2023/12/28 00:37:59] [ info] [input] pausing dummy.0
fluent-bit -i dummy -o exit -f 30  0.01s user 0.01s system 0% cpu 31.132 total
```

**Comment by joshsleeper** [comment]
dear stale-bot, nah.

**Comment by edsiper** [comment]
Since the grace period only works when  there is still pending data, is the suggestion to have an additional option to enforce the grace period respect the time set ? 

what would be the ideal behavior of another option that sets the wait period if a grace period already exists ?

**Comment by Benjaminvdv** [comment]
> Since the grace period only works when there is still pending data, is the suggestion to have an additional option to enforce the grace period respect the time set ?

That could be an option, yes. The OPA project formulates it differently though:
```
      --shutdown-grace-period int      set the time (in seconds) that the server will wait to gracefully shut down. This flag is valid in 'e2e' mode only. (default 10)
      --shutdown-wait-period int       set the time (in seconds) that the server will wait before initiating shutdown. This flag is valid in 'e2e' mode only.
```
So instead of enforcing the grace period, it simple waits before initiating the (graceful) shutdown.

> what would be the ideal behavior of another option that sets the wait period if a grace period already exists ?

Ideally it would make sure that the shutdown waits for a configured (minimum) amount of time so that it gives time for the (main) application to shutdown as well.

**Comment by vkras** [comment]
Our current workaround (compatible with multi-arch builds) is following:
1. Add sleep.c into docker project
```
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

int main(int argc, char *argv[])
{
    // Check if the correct number of arguments is provided
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <seconds>\n", argv[0]);
        return 1;
    }

    // Convert the argument to an integer
    int seconds = atoi(argv[1]);

    // Check if the provided argument is a positive integer
    if (seconds <= 0) {
        fprintf(stderr, "Error: Please provide a positive number of seconds.\n");
        return 1;
    }

    // Sleep for the specified number of seconds
    sleep(seconds);

    return 0;
}
```

2. Create Dockerfile for the side car: 
```
# Stage 1: Build the C program
FROM gcc:latest AS build

WORKDIR /app

# Copy the C source file into the container
COPY sleep.c .

# Compile the C program, static with no external dependencies
RUN gcc -static -o sleep sleep.c

# Make the binary executable
RUN chmod 555 sleep

# Stage 2: Non-debug fluent-bit image with sleep command copied inside
FROM fluent/fluent-bit:3.0

# Copy the compiled binary from the build stage
COPY --from=build /app/sleep /bin/sleep

# keep the original entry point
```
3. Add preStop hook to fluent-bit side car: 
```
       lifecycle:
          preStop:
            exec:
              command:
              - /bin/sleep
              - "5"
```

**Comment by joshsleeper** [comment]
for those still interested, I _think_ the behavior change added in #9952 and released in `4.0.3` might help somewhat?
